### PR TITLE
[bridge] Set stoppedTime while workspace is stopping

### LINKED ISSUE #4818: Revise setting workspace stop time: Don't let workspace stopping time eat up users' hours
Currently, the time during the workspace stopping phase decreases the users' hours. When the workspace is stuck in stopping, that could eat up users' hours completely. Reconsider setting the stopping time that calculates the hours as soon as the stopping phase begins.

/cc @svenefftinge @csweichel

### PR DESCRIPTION
**Pull request by csweichel**
there's no need to include the time our workspaces take to stop into account when computing workspace runtime (which is accounting relevant).

fixes #4818

### CODE REVIEW

**Comment by geropl** [comment]
Code-wise this looks ok.

But to me this feels like accumulating debt: Why not have a 2nd field called `stoppingTime`? :thinking:

**Comment by csweichel** [comment]
> Code-wise this looks ok.
> 
> But to me this feels like accumulating debt: Why not have a 2nd field called `stoppingTime`? 🤔

done

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4833#" title="Author self-approved">csweichel</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **geropl** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @geropl` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by geropl** [comment]
> done

Awesome. Will start review now...

File: components/ws-manager-bridge/src/bridge.ts [hunk 1]
```diff
@@ -219,10 +219,20 @@ export class WorkspaceManagerBridge implements Disposable {
...
+                    if (!instance.stoppingTime) {
+                        // The first time a workspace enters stopping we record that as it's stoppingTime time.
+                        // This way we don't take the time a workspace requires to stop into account when
+                        // computing the time a workspace instance was running.
+                        instance.stoppingTime = new Date().toISOString();
+                    }
                     break;
                 case WorkspacePhase.STOPPED:
-                    instance.stoppedTime = new Date().toISOString();
                     instance.status.phase = "stopped";
```
Comment (geropl) [inline components/ws-manager-bridge/src/bridge.ts:230]: `instance.stoppedTime = new Date().toISOString();` went missing here, should be re-added.

**Review by geropl — CHANGES_REQUESTED** [review]
Looks good! IMO there two things missing:
 1. selecting `stoppingTime` when selecting sessions from the DB
 1. setting `stoppedTime` in bridge
 
 I made the changes [here](https://github.com/gitpod-io/gitpod/commit/3f1eefc1ee54e580295e419ea927323d041aca06), feel free to cherry-pick.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4833#" title="Author self-approved">csweichel</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-db/src/typeorm/entity/db-workspace-instance.ts [hunk 2]
```diff
@@ -40,6 +40,12 @@ export class DBWorkspaceInstance implements WorkspaceInstance {
     })
     deployedTime?: string;
 
+    @Column({
+        default: '',
+        transformer: Transformer.MAP_EMPTY_STR_TO_UNDEFINED
+    })
+    stoppingTime?: string;
```
Comment (svenefftinge) [inline components/gitpod-db/src/typeorm/entity/db-workspace-instance.ts:47]: a comment that explains what the difference between `stoppingTime` and `stoppedTime` is would help avoid missunderstandings.

**Comment by csweichel** [comment]
Thank you @geropl - I've cherry-picked your changes.

File: components/server/src/workspace/workspace-starter.ts [hunk 3]
```diff
@@ -222,6 +222,7 @@ export class WorkspaceStarter {
             // We may have never actually started the workspace which means that ws-manager-bridge never set a workspace status.
             // We have to set that status ourselves.
             instance.status.phase = 'stopped';
+            instance.stoppingTime = new Date().toISOString();
```
Comment (geropl) [inline components/server/src/workspace/workspace-starter.ts:225]: Maybe a nit, but: guaranteed `stoppingTime == stoppedTime` is a nice thing to have for downstream usage: analytics, bug hunting, ad-hoc DB fixes, etc.

File: components/ws-manager-bridge/src/bridge.ts [hunk 4]
```diff
@@ -299,6 +311,7 @@ export class WorkspaceManagerBridge implements Disposable {
 
             log.info({instanceId, workspaceId: instance.workspaceId}, "Database says the instance is starting for too long or running, but wsman does not know about it. Marking as stopped in database.", {installation});
             instance.status.phase = "stopped";
+            instance.stoppingTime = new Date().toISOString();
```
Comment (geropl) [inline components/ws-manager-bridge/src/bridge.ts:314]: Maybe a nit, but: guaranteed `stoppingTime == stoppedTime` is a nice thing to have for downstream usage: analytics, bug hunting, ad-hoc DB fixes, etc.

**Comment by geropl** [comment]
I just tested and somehow ended up in a re-start loop :confused: Will investigate.

**Review by geropl — APPROVED** [review]
Works as advertised! Feels wholesome to have that column, finally :-D

Happy to fix the nits separately (if at all).

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 564c438d498ff7b96eaa7fceed0323b3463b244f</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4833#" title="Author self-approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4833#pullrequestreview-710372785" title="Approved">geropl</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [geropl]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**